src/Halogen/Canvas/Interact.purs

```diff
@@ -0,0 +1,128 @@
+module Halogen.Canvas.Interact
+  ( InputEvent(..)
+  , stopInputEventPropagation
+  , component
+  ) where
+
+import Prelude
+
+import Control.Monad.Reader (runReaderT)
+import Control.Monad.Rec.Class (class MonadRec)
+import Data.Maybe (Maybe(..))
+import Data.Traversable (traverse_)
+import Effect (Effect)
+import Effect.Aff.Class (class MonadAff)
+import Graphics.Canvas (getContext2D)
+import Graphics.Canvas.Free (CanvasContext, CanvasT, clearRect, getHeight, getWidth, runCanvasT, withContext)
+import Halogen as H
+import Halogen.HTML as HH
+import Halogen.HTML.Events as HE
+import Halogen.HTML.Properties as HP
+import Unsafe.Coerce (unsafeCoerce)
+import Web.DOM.Element (DOMRect, getBoundingClientRect)
+import Web.Event.Event (stopPropagation)
+import Web.UIEvent.KeyboardEvent (KeyboardEvent)
+import Web.UIEvent.KeyboardEvent as KeyboardEvent
+import Web.UIEvent.MouseEvent (MouseEvent)
+import Web.UIEvent.MouseEvent as MouseEvent
+
+type Dimensions =
+  { width :: Int
+  , height :: Int
+  }
+
+data InputEvent =
+    KeyDown KeyboardEvent
+  | KeyUp KeyboardEvent
+  | Click MouseEvent 
+  | DoubleClick MouseEvent
+  | MouseDown MouseEvent
+  | MouseUp MouseEvent
+  | MouseEnter MouseEvent
+  | MouseLeave MouseEvent
+  | MouseMove MouseEvent
+
+type Interact world m =
+  { dimensions :: Dimensions 
+  , world :: world
+  , draw :: world -> CanvasT m Unit
+  , input :: InputEvent -> DOMRect -> world -> world
+  }
```

Comment:
Interact is a component with a `world` state that has a pure transition function responding to input. This is similar to the `Game` component but without the `requestAnimationFrame` loop which requires significant compute overhead even when no animation is being done. The notion is that this is [similar to a Gloss event loop except the picture is drawn in an imperative style in `CanvasT`](https://hackage.haskell.org/package/gloss-1.13.2.2/docs/Graphics-Gloss.html).